File: cat/cat.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    // If no file is provided, read from standard input
    if (argc < 2) {
        char buffer[BUFSIZ];
        size_t n;
        while ((n = fread(buffer, 1, BUFSIZ, stdin)) > 0) {
            fwrite(buffer, 1, n, stdout);
        }
    } else {
        // Iterate over each file provided as an argument
        for (int i = 1; i < argc; i++) {
            FILE *fp = fopen(argv[i], "r");
            if (fp == NULL) {
                perror(argv[i]);
                continue;
            }
            char buffer[BUFSIZ];
            size_t n;
            while ((n = fread(buffer, 1, BUFSIZ, fp)) > 0) {
                fwrite(buffer, 1, n, stdout);
            }
            fclose(fp);
        }
    }
    return 0;
}

File: cat/cat.rs
use std::env;
use std::fs::File;
use std::io::{self, Read, Write};

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();

    // If no files are provided, read from standard input
    if args.len() < 2 {
        let mut buffer = String::new();
        io::stdin().read_to_string(&mut buffer)?;
        print!("{}", buffer);
    } else {
        // Iterate over each file provided as an argument
        for filename in &args[1..] {
            let mut file = match File::open(filename) {
                Ok(f) => f,
                Err(e) => {
                    eprintln!("Error opening {}: {}", filename, e);
                    continue;
                }
            };

            let mut contents = String::new();
            if let Err(e) = file.read_to_string(&mut contents) {
                eprintln!("Error reading {}: {}", filename, e);
                continue;
            }
            print!("{}", contents);
        }
    }
    io::stdout().flush()?;
    Ok(())
}
